{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "105744c5-15f9-4112-8ea9-7f8295aec581",
   "metadata": {},
   "source": [
    "# Errors and Exceptions\n",
    "teaching: 30\n",
    "exercises: 0\n",
    "questions:\n",
    "- \"How does Python report errors?\"\n",
    "- \"How can I handle errors in Python programs?\"\n",
    "objectives:\n",
    "- \"To be able to read a traceback, and determine where the error took place and what type it is.\"\n",
    "- \"To be able to describe the types of situations in which syntax errors,\n",
    "   indentation errors, name errors, index errors, and missing file errors occur.\"\n",
    "keypoints:\n",
    "- \"Tracebacks can look intimidating, but they give us a lot of useful information about\n",
    "   what went wrong in our program, including where the error occurred and\n",
    "   what type of error it was.\"\n",
    "- \"An error having to do with the 'grammar' or syntax of the program is called a `SyntaxError`.\n",
    "   If the issue has to do with how the code is indented,\n",
    "   then it will be called an `IndentationError`.\"\n",
    "- \"A `NameError` will occur when trying to use a variable that does not exist. Possible causes are\n",
    "  that a variable definition is missing, a variable reference differs from its definition\n",
    "  in spelling or capitalization, or the code contains a string that is missing quotes around it.\"\n",
    "- \"Containers like lists and strings will generate errors if you try to access items\n",
    "   in them that do not exist. This type of error is called an `IndexError`.\"\n",
    "- \"Trying to read a file that does not exist will give you an `FileNotFoundError`.\n",
    "   Trying to read a file that is open for writing, or writing to a file that is open for reading,\n",
    "   will give you an `IOError`.\"\n",
    "---\n",
    "\n",
    "Every programmer encounters errors,\n",
    "both those who are just beginning,\n",
    "and those who have been programming for years.\n",
    "Encountering errors and exceptions can be very frustrating at times,\n",
    "and can make coding feel like a hopeless endeavour.\n",
    "However,\n",
    "understanding what the different types of errors are\n",
    "and when you are likely to encounter them can help a lot.\n",
    "Once you know *why* you get certain types of errors,\n",
    "they become much easier to fix.\n",
    "\n",
    "Errors in Python have a very specific form,\n",
    "called a __traceback__.\n",
    "Let's examine one:\n",
    "\n",
    "```Python\n",
    "# This code has an intentional error. You can type it directly or\n",
    "# use it for reference to understand the error message below.\n",
    "def favorite_ice_cream():\n",
    "    ice_creams = [\n",
    "        'chocolate',\n",
    "        'vanilla',\n",
    "        'strawberry'\n",
    "    ]\n",
    "    print(ice_creams[3])\n",
    "\n",
    "favorite_ice_cream()\n",
    "```\n",
    "\n",
    "\n",
    "This particular traceback has two levels.\n",
    "You can determine the number of levels by looking for the number of arrows on the left hand side.\n",
    "In this case:\n",
    "\n",
    "1.  The first shows code from the cell above,\n",
    "    with an arrow pointing to Line 11 (which is `favorite_ice_cream()`).\n",
    "\n",
    "2.  The second shows some code in the function `favorite_ice_cream`,\n",
    "    with an arrow pointing to Line 9 (which is `print(ice_creams[3])`).\n",
    "\n",
    "The last level is the actual place where the error occurred.\n",
    "The other level(s) show what function the program executed to get to the next level down.\n",
    "So, in this case, the program first performed a\n",
    "__function call__ to the function `favorite_ice_cream`.\n",
    "Inside this function,\n",
    "the program encountered an error on Line 6, when it tried to run the code `print(ice_creams[3])`.\n",
    "\n",
    "So what error did the program actually encounter?\n",
    "In the last line of the traceback,\n",
    "Python helpfully tells us the category or type of error (in this case, it is an `IndexError`)\n",
    "and a more detailed error message (in this case, it says \"list index out of range\").\n",
    "\n",
    "If you encounter an error and don't know what it means,\n",
    "it is still important to read the traceback closely.\n",
    "That way,\n",
    "if you fix the error,\n",
    "but encounter a new one,\n",
    "you can tell that the error changed.\n",
    "Additionally,\n",
    "sometimes knowing *where* the error occurred is enough to fix it,\n",
    "even if you don't entirely understand the message.\n",
    "\n",
    "If you do encounter an error you don't recognize,\n",
    "try looking at the\n",
    "[official documentation on errors](http://docs.python.org/3/library/exceptions.html).\n",
    "However,\n",
    "note that you may not always be able to find the error there,\n",
    "as it is possible to create custom errors.\n",
    "In that case,\n",
    "hopefully the custom error message is informative enough to help you figure out what went wrong.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eee84e1-df69-4060-a985-402cc7f939f3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1bc066d4-7953-4dee-b344-1873be7f938c",
   "metadata": {},
   "source": [
    "## Syntax Errors\n",
    "\n",
    "When you forget a colon at the end of a line,\n",
    "accidentally add one space too many when indenting under an `if` statement,\n",
    "or forget a parenthesis,\n",
    "you will encounter a [syntax error]({{ page.root }}/reference.html#syntax-error).\n",
    "This means that Python couldn't figure out how to read your program.\n",
    "This is similar to forgetting punctuation in English:\n",
    "for example,\n",
    "this text is difficult to read there is no punctuation there is also no capitalization\n",
    "why is this hard because you have to figure out where each sentence ends\n",
    "you also have to figure out where each sentence begins\n",
    "to some extent it might be ambiguous if there should be a sentence break or not\n",
    "\n",
    "People can typically figure out what is meant by text with no punctuation,\n",
    "but people are much smarter than computers.\n",
    "If Python doesn't know how to read the program,\n",
    "it will give up and inform you with an error.\n",
    "For example:\n",
    "\n",
    "```Python\n",
    "def some_function()\n",
    "    msg = 'hello, world!'\n",
    "    print(msg)\n",
    "     return msg\n",
    "```\n",
    "\n",
    "Here, Python tells us that there is a `SyntaxError` on line 1,\n",
    "and even puts a little arrow in the place where there is an issue.\n",
    "In this case the problem is that the function definition is missing a colon at the end.\n",
    "\n",
    "Actually, the function above has *two* issues with syntax.\n",
    "If we fix the problem with the colon,\n",
    "we see that there is *also* an `IndentationError`,\n",
    "which means that the lines in the function definition do not all have the same indentation:\n",
    "\n",
    "```Python\n",
    "def some_function():\n",
    "    msg = 'hello, world!'\n",
    "    print(msg)\n",
    "     return msg\n",
    "```\n",
    "\n",
    "Both `SyntaxError` and `IndentationError` indicate a problem with the syntax of your program,\n",
    "but an `IndentationError` is more specific:\n",
    "it *always* means that there is a problem with how your code is indented.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b60c09c-71e2-4c10-bfde-0869281fbc68",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c130d82d-379e-42a2-b534-1b1c9e79d609",
   "metadata": {},
   "source": [
    "> ## Tabs and Spaces\n",
    ">\n",
    "> Some indentation errors are harder to spot than others.\n",
    "> In particular, mixing spaces and tabs can be difficult to spot\n",
    "> because they are both __whitespace__.\n",
    "> In the example below, the first two lines in the body of the function\n",
    "> `some_function` are indented with tabs, while the third line &mdash; with spaces.\n",
    "> If you're working in a Jupyter notebook, be sure to copy and paste this example\n",
    "> rather than trying to type it in manually because Jupyter automatically replaces\n",
    "> tabs with spaces.\n",
    ">\n",
    "> ```Python\n",
    "> def some_function():\n",
    "> \tmsg = 'hello, world!'\n",
    "> \tprint(msg)\n",
    ">         return msg\n",
    "> ```\n",
    "> \n",
    ">\n",
    "> Visually it is impossible to spot the error.\n",
    "> Fortunately, Python does not allow you to mix tabs and spaces.\n",
    ">\n",
    "> ```\n",
    ">   File \"<ipython-input-5-653b36fbcd41>\", line 4\n",
    ">     return msg\n",
    ">               ^\n",
    "> TabError: inconsistent use of tabs and spaces in indentation\n",
    "> ```\n",
    ">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3f9f9d0-6556-4096-9e91-de4875d82f44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e81105a4-7643-4ed6-b26a-390b41c6a802",
   "metadata": {},
   "source": [
    "## Variable Name Errors\n",
    "\n",
    "Another very common type of error is called a `NameError`,\n",
    "and occurs when you try to use a variable that does not exist.\n",
    "For example:\n",
    "\n",
    "```Python\n",
    "print(a)\n",
    "```\n",
    "\n",
    "Variable name errors come with some of the most informative error messages,\n",
    "which are usually of the form \"name 'the_variable_name' is not defined\".\n",
    "\n",
    "Why does this error message occur?\n",
    "That's a harder question to answer,\n",
    "because it depends on what your code is supposed to do.\n",
    "However,\n",
    "there are a few very common reasons why you might have an undefined variable.\n",
    "The first is that you meant to use a\n",
    "__string__, but forgot to put quotes around it:\n",
    "\n",
    "```Python\n",
    "print(hello)\n",
    "```\n",
    "\n",
    "The second reason is that you might be trying to use a variable that does not yet exist.\n",
    "In the following example,\n",
    "`count` should have been defined (e.g., with `count = 0`) before the for loop:\n",
    "\n",
    "```Python\n",
    "for number in range(10):\n",
    "    count = count + number\n",
    "print('The count is:', count)\n",
    "```\n",
    "\n",
    "Finally, the third possibility is that you made a typo when you were writing your code.\n",
    "Let's say we fixed the error above by adding the line `Count = 0` before the for loop.\n",
    "Frustratingly, this actually does not fix the error.\n",
    "Remember that variables are [case-sensitive]({{ page.root }}/reference.html#case-sensitive),\n",
    "so the variable `count` is different from `Count`. We still get the same error,\n",
    "because we still have not defined `count`:\n",
    "\n",
    "```Python\n",
    "Count = 0\n",
    "for number in range(10):\n",
    "    count = count + number\n",
    "print('The count is:', count)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ee1fa40-3936-4b9f-94c3-0771b9e4c79c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "aa679e9a-8250-4f96-9849-09cb20f864e0",
   "metadata": {},
   "source": [
    "## Index Errors\n",
    "\n",
    "Next up are errors having to do with containers (like lists and strings) and the items within them.\n",
    "If you try to access an item in a list or a string that does not exist,\n",
    "then you will get an error.\n",
    "This makes sense:\n",
    "if you asked someone what day they would like to get coffee,\n",
    "and they answered \"caturday\",\n",
    "you might be a bit annoyed.\n",
    "Python gets similarly annoyed if you try to ask it for an item that doesn't exist:\n",
    "\n",
    "```Python\n",
    "letters = ['a', 'b', 'c']\n",
    "print('Letter #1 is', letters[0])\n",
    "print('Letter #2 is', letters[1])\n",
    "print('Letter #3 is', letters[2])\n",
    "print('Letter #4 is', letters[3])\n",
    "```\n",
    "\n",
    "Here,\n",
    "Python is telling us that there is an `IndexError` in our code,\n",
    "meaning we tried to access a list index that did not exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "478ff304-4b84-4229-9134-1b0e20f1c2e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0ec52ad5-6700-4ac4-8dbf-62499f904d6c",
   "metadata": {},
   "source": [
    "## File Errors\n",
    "\n",
    "The last type of error we'll cover today\n",
    "are those associated with reading and writing files: `FileNotFoundError`.\n",
    "If you try to read a file that does not exist,\n",
    "you will receive a `FileNotFoundError` telling you so.\n",
    "If you attempt to write to a file that was opened read-only, Python 3\n",
    "returns an `UnsupportedOperationError`.\n",
    "More generally, problems with input and output manifest as\n",
    "`IOError`s or `OSError`s, depending on the version of Python you use.\n",
    "\n",
    "```Python\n",
    "file_handle = open('myfile.txt', 'r')\n",
    "```\n",
    "\n",
    "One reason for receiving this error is that you specified an incorrect path to the file.\n",
    "For example,\n",
    "if I am currently in a folder called `myproject`,\n",
    "and I have a file in `myproject/writing/myfile.txt`,\n",
    "but I try to open `myfile.txt`,\n",
    "this will fail.\n",
    "The correct path would be `writing/myfile.txt`.\n",
    "It is also possible that the file name or its path contains a typo.\n",
    "\n",
    "A related issue can occur if you use the \"read\" flag instead of the \"write\" flag.\n",
    "Python will not give you an error if you try to open a file for writing\n",
    "when the file does not exist.\n",
    "However,\n",
    "if you meant to open a file for reading,\n",
    "but accidentally opened it for writing,\n",
    "and then try to read from it,\n",
    "you will get an `UnsupportedOperation` error\n",
    "telling you that the file was not opened for reading:\n",
    "\n",
    "```Python\n",
    "file_handle = open('myfile.txt', 'w')\n",
    "file_handle.read()\n",
    "```\n",
    "\n",
    "These are the most common errors with files,\n",
    "though many others exist.\n",
    "If you get an error that you've never seen before,\n",
    "searching the Internet for that error type\n",
    "often reveals common reasons why you might get that error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7f5e1db-3ca7-43e4-bf99-652ed0f809fe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
